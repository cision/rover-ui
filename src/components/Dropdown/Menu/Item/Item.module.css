.Item {
  text-align: left;
  display: block;
  width: 100%;
  border-radius: 0;
}

.button:hover:not(.disabled),
.button:active:not(.disabled) {
  background-color: var(--rvr-gray-10);
}

.content {
  cursor: default;
}
