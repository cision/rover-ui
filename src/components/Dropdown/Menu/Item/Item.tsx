import React from 'react';
import classNames from 'classnames';

import Button from '../../../Button';

import type {
  AnchorElementProps,
  ButtonElementProps,
} from '../../../Button/Button';

import styles from './Item.module.css';

interface DivItemProps extends React.AllHTMLAttributes<HTMLDivElement> {
  disabled?: boolean;
  forwardedRef?: React.Ref<HTMLDivElement>;
}
interface AnchorItemProps extends AnchorElementProps {
  forwardedRef?: React.Ref<HTMLAnchorElement>;
}

interface ButtonItemProps extends ButtonElementProps {
  forwardedRef?: React.Ref<HTMLButtonElement>;
}

export type ItemProps = ButtonItemProps | AnchorItemProps | DivItemProps;

// Guard to check if href exists in props
const hasHref = (props: ItemProps): props is AnchorElementProps =>
  'href' in props && props.href !== undefined;

const Item: React.FC<ItemProps> = ({
  className = '',
  forwardedRef,
  ...passedProps
}) => {
  /**
   * If an `href` prop is passed, the rendered element automatically renders
   * as a bare <a> element with default styles
   */
  if (hasHref(passedProps) || passedProps.onClick) {
    return (
      // eslint-disable-next-line jsx-a11y/anchor-has-content
      <Button.WithRef
        className={classNames(styles.Item, styles.button, className, {
          [styles.disabled]: passedProps.disabled,
        })}
        level="text"
        ref={forwardedRef}
        size="md"
        {...(passedProps as ButtonElementProps)}
      />
    );
  }

  /**
   * If a no `href` or `onClick` props are passed, the rendered element simply renders
   * as a <div> element with some default styles
   */
  return (
    <Button.WithRef
      className={classNames(styles.Item, styles.content, className)}
      level="text"
      ref={forwardedRef}
      size="md"
      tag="div"
      {...(passedProps as ButtonElementProps)}
    />
  );
};

// Temporary
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export default React.forwardRef<any, any>((props, ref) => (
  <Item {...props} forwardedRef={ref} />
));
