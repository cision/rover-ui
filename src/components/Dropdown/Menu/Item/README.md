# \<Dropdown.Menu.Item\>

This dropdown menu item should _always_ be a direct child of a `Dropdown.Menu` component.

It adds consistent padding and wraps menu items in either a `<span>` or `<button>` depending on whether or not there's an `onClick` property provided.
