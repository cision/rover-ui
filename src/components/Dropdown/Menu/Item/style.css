.Item {
  background: transparent none;
  padding: var(--rvr-space-sm) var(--rvr-space-lg);
  -webkit-appearance: none;
  border: 0 none transparent;
  transition: 100ms var(--rvr-easeInOutQuad) background-color, 100ms var(--rvr-easeInOutQuad) color;
  text-align: left;
  display: block;
  box-sizing: border-box;
  width: 100%;
  font-family: var(--rvr-base-font-family);
  font-size: var(--rvr-font-size-md);
  line-height: var(--rvr-line-height-sm);
  color: var(--rvr-gray-90);
}

.button {
  cursor: pointer;
  color: var(--rvr-gray-90);
}

.button:hover,
.button:active {
  background-color: var(--rvr-gray-lite-5);
}

.link {
  cursor: pointer;
  color: var(--rvr-gray-90);
}

.link:hover,
.link:active {
  background-color: var(--rvr-gray-lite-5);
}
