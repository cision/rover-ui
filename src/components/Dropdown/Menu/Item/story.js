import React from 'react';
import { storiesOf } from '@storybook/react';
import { select, text } from '@storybook/addon-knobs';
import { action } from '@storybook/addon-actions';

import Media from '../../../Media';

import Item from './';
import Readme from './README.md';

storiesOf('Star Systems/Dropdown/Menu/Item', module)
  .addParameters({
    readme: {
      sidebar: Readme,
    },
  })
  .add('Overview', () => (
    <div
      style={{
        margin: '0 auto',
        width: '300px',
      }}
    >
      <Item
        onClick={
          select('onClick', ['() => {}', 'undefined'], '() => {}') ===
          'undefined'
            ? undefined
            : () => {}
        }
      >
        {text('children', 'Dropdown menu item text')}
      </Item>
    </div>
  ))
  .add(
    'Examples',
    () => (
      <div
        style={{
          margin: '0 auto',
          width: '300px',
          background: 'white',
        }}
      >
        <Item>With basic text content, no onClick</Item>
        <Item onClick={action('onClick')}>
          With basic text content, and onClick
        </Item>
        <Item>
          <Media>
            <Media.Item style={{ paddingRight: '10px' }}>
              <img
                alt="Random from Unsplash"
                style={{ borderRadius: '20px' }}
                src="https://source.unsplash.com/random/40x40"
              />
            </Media.Item>
            <Media.Body>
              <span>With some React node content, no onClick</span>
            </Media.Body>
          </Media>
        </Item>
        <Item onClick={action('onClick')}>
          <Media>
            <Media.Item style={{ paddingRight: '10px' }}>
              <img
                alt="Random from Unsplash"
                style={{ borderRadius: '20px' }}
                src="https://source.unsplash.com/random/40x40"
              />
            </Media.Item>
            <Media.Body>
              <span>With some React node content, and onClick</span>
            </Media.Body>
          </Media>
        </Item>
      </div>
    ),
    {
      info: {
        inline: true,
        source: true,
      },
    }
  );
