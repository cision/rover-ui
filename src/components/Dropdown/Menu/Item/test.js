import React from 'react';
import { shallow } from 'enzyme';

import Item from './';

describe('Item', () => {
  it('renders', () => {
    shallow(<Item>Boom</Item>);
  });

  it('renders content', () => {
    const wrapper = shallow(<Item>Boom</Item>);
    expect(wrapper.text()).toEqual('Boom');
  });

  describe("when there's an href", () => {
    it('renders a link', () => {
      const wrapper = shallow(<Item href="foo">Hi</Item>);
      expect(wrapper.find('a').length).toEqual(1);
    });

    describe('and an onClick', () => {
      it('still renders a link', () => {
        const wrapper = shallow(
          <Item href="foo" onClick={() => {}}>
            Hi
          </Item>
        );
        expect(wrapper.find('a').length).toEqual(1);
        expect(wrapper.find('button').length).toEqual(0);
      });
    });
  });

  describe("when there's an onClick", () => {
    const wrapper = shallow(<Item onClick={() => {}}>Hi</Item>);
    expect(wrapper.find('button').length).toEqual(1);
  });

  describe('when children is a React node', () => {
    it('still renders', () => {
      const wrapper = shallow(
        <Item>
          <span>Span!</span>
        </Item>
      );
      expect(wrapper.find('span').length).toEqual(1);
      expect(wrapper.text()).toEqual('Span!');
    });
  });
});
