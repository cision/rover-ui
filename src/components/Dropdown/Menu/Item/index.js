import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import style from './style.css';

const Item = ({ className, ...passedProps }) => {
  if (passedProps.href) {
    // The anchor's content is provided by passedProps
    /* eslint-disable jsx-a11y/anchor-has-content */
    return (
      <a
        className={classNames(style.Item, style.link, className)}
        {...passedProps}
      />
    );
    /* eslint-enable jsx-a11y/anchor-has-content */
  }

  if (passedProps.onClick) {
    return (
      <button
        className={classNames(style.Item, style.button, className)}
        type="button"
        {...passedProps}
      />
    );
  }

  return (
    <div
      className={classNames(style.Item, style.content, className)}
      {...passedProps}
    />
  );
};

Item.propTypes = {
  className: PropTypes.string,
  href: PropTypes.string,
  onClick: PropTypes.func,
};

Item.defaultProps = {
  className: '',
  href: undefined,
  onClick: undefined,
};

export default Item;
