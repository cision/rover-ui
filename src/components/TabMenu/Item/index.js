import React from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';

import style from './Item.module.css';

const Item = ({ className: customClassName, active, children, ...props }) => {
  const isChildActive = (child) => {
    if (React.isValidElement(child)) {
      return child.props.className.indexOf('active') >= 0;
    }
    return false;
  };

  const className = classNames(style.Item, customClassName, {
    [style.active]: active || isChildActive(children),
  });
  return (
    <li className={className} {...props}>
      {children}
    </li>
  );
};

Item.propTypes = {
  className: PropTypes.string,
  active: PropTypes.bool,
  children: PropTypes.node.isRequired,
  onClick: PropTypes.func,
};

Item.defaultProps = {
  className: '',
  active: false,
  onClick: () => {},
};

export default Item;
