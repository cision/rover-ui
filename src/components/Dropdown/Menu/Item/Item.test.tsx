import React from 'react';
import { render } from '@testing-library/react';

import Item from '.';

describe('Item', () => {
  it('renders', () => {
    const { baseElement } = render(<Item>Boom</Item>);
    expect(baseElement).toMatchSnapshot();
  });

  it('renders content', () => {
    const { getByText } = render(<Item>Boom</Item>);
    expect(getByText('Boom')).toBeTruthy();
  });

  describe("when there's an href", () => {
    it('renders a link', () => {
      const { getByText } = render(<Item href="foo">Hi</Item>);
      const menuItem = getByText('Hi') as HTMLAnchorElement;
      expect(menuItem.tagName).toEqual('A');
      expect(menuItem.href).toEqual('http://localhost/foo');
    });

    describe('and an onClick', () => {
      it('still renders a link', () => {
        const { getByText } = render(
          <Item href="foo" onClick={() => {}}>
            Hi
          </Item>
        );

        const menuItem = getByText('Hi');
        expect(menuItem.tagName).toEqual('A');
      });
    });
  });

  describe("when there's an onClick", () => {
    const { getByText } = render(<Item onClick={() => {}}>Hi</Item>);

    const menuItem = getByText('Hi');
    expect(menuItem.tagName).toEqual('BUTTON');
  });

  describe('when children is a React node', () => {
    it('still renders', () => {
      const { baseElement, getByText } = render(
        <Item>
          <span>Span!</span>
        </Item>
      );
      expect(baseElement).toMatchSnapshot();
      expect(getByText('Span!')).toBeTruthy();
    });
  });
});
