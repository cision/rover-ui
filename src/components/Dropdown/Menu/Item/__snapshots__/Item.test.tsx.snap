// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item renders 1`] = `
<body>
  <div>
    <button
      class="Item button Button text md"
      type="button"
    >
      Hi
    </button>
  </div>
  <div>
    <div
      class="Item content Button text md"
    >
      Boom
    </div>
  </div>
</body>
`;

exports[`Item when children is a React node still renders 1`] = `
<body>
  <div>
    <div
      class="Item content Button text md"
    >
      <span>
        Span!
      </span>
    </div>
  </div>
</body>
`;
